NeoML/src/Dnn/Layers/MatrixMultiplicationLayer.cpp

```diff
@@ -31,8 +35,35 @@ void CMatrixMultiplicationLayer::Serialize( CArchive& archive )
 {
 	archive.SerializeVersion( MatrixMultiplicationLayerVersion );
 	CBaseLayer::Serialize( archive );
+	if( archive.IsLoading() ) {
+		destroySmallMatricesMulDesc();
+	}
 }
 
+const CSmallMatricesMultiplyDesc* CMatrixMultiplicationLayer::initSmallMatricesMulDesc(
+	TSMMD type, int firstHeight, int firstWidth, int secondWidth, int resultWidth )
+{
+	NeoPresume( inputBlobs[0] != nullptr || inputDiffBlobs[0] != nullptr );
+	NeoPresume( outputBlobs[0] != nullptr || outputDiffBlobs[0] != nullptr );
+
+	if( smallMatricesMulDescs[type].FaultFlag == false && smallMatricesMulDescs[type].Ptr == nullptr ) {
+		smallMatricesMulDescs[type].Ptr = MathEngine().InitSmallMatricesMultiplyDesc( firstHeight, firstWidth, secondWidth, resultWidth,
+			/*resultAdd*/false, /*trans2*/( type == TSMMD_Backward ), /*trans1*/( type == TSMMD_SecondBackward ) );
+		smallMatricesMulDescs[type].FaultFlag = (  smallMatricesMulDescs[type].Ptr == nullptr ); //fault to create JIT
+	}
+	return smallMatricesMulDescs[type].Ptr;
+}
+
+void CMatrixMultiplicationLayer::destroySmallMatricesMulDesc()
+{
+	for( auto& desc : smallMatricesMulDescs ) {
+		if( desc.Ptr != nullptr ) {
+			delete desc.Ptr;
+			desc.Ptr = nullptr;
+			desc.FaultFlag = false;
```

Comment:
The reshape may lead to change in blob sizes.

Let's imagine the situation:
First `CDnn::Reshape()`, bad sizes which can't be handled by JIT (`desc.Ptr == nullptr && desc.FaultFlag == true)
Second `CDnn::Reshape()` good sizes, but the state after the first won't be reset!

Reply:
fixed